strings/knuth_morris_pratt.cpp

```diff
@@ -18,78 +18,84 @@
 #else
 #include <cstring>
 #endif
+#include <cassert>
 #include <vector>
 
-/** \namespace string_search
- * \brief String search algorithms
+/**
+ * @namespace string_search
+ * @brief String search algorithms
  */
 namespace string_search {
 /**
- * Generate the partial match table aka failure function for a pattern to
+ * @brief Generate the partial match table aka failure function for a pattern to
  * search.
- * \param[in] pattern text for which to create the partial match table
- * \returns the partial match table as a vector array
+ * @param pattern text for which to create the partial match table
+ * @returns the partial match table as a vector array
  */
-std::vector<int> getFailureArray(const std::string &pattern) {
-    int pattern_length = pattern.size();
-    std::vector<int> failure(pattern_length + 1);
-    failure[0] = -1;
-    int j = -1;
-
+std::vector<size_t> getFailureArray(const std::string &pattern) {
+    size_t pattern_length = pattern.size();
+    std::vector<size_t> failure(pattern_length + 1);
+    failure[0] = std::string::npos;
+    size_t j = std::string::npos;
     for (int i = 0; i < pattern_length; i++) {
-        while (j != -1 && pattern[j] != pattern[i]) {
+        while (j != std::string::npos && pattern[j] != pattern[i]) {
             j = failure[j];
         }
-        j++;
-        failure[i + 1] = j;
+        failure[i + 1] = ++j;
     }
     return failure;
 }
 
 /**
- * KMP algorithm to find a pattern in a text
- * \param[in] pattern string pattern to search
- * \param[in] text text in which to search
- * \returns `true` if pattern was found
- * \returns `false` if pattern was not found
+ * @brief KMP algorithm to find a pattern in a text
+ * @param pattern string pattern to search
+ * @param text text in which to search
+ * @returns the starting index of the pattern if found
+ * @returns `std::string::npos` if not found
  */
-bool kmp(const std::string &pattern, const std::string &text) {
-    int text_length = text.size(), pattern_length = pattern.size();
-    std::vector<int> failure = getFailureArray(pattern);
-
-    int k = 0;
-    for (int j = 0; j < text_length; j++) {
-        while (k != -1 && pattern[k] != text[j]) {
+size_t kmp(const std::string &pattern, const std::string &text) {
+    if (pattern.empty()) {
+        return 0;
```

Comment:
Shouldn't this be npos now?

```suggestion
        return std::string::npos;
```

Reply:
Personally, I think it can be either `0` or `std::string::npos`, depending on different use cases.

Reply:
and I test the following codes, it print `0`.
```cpp
#include <iostream>

int main() {
    std::string str = "aaa";
    std::cout << str.find("") << std::endl;
    return 0;
}
```

Reply:
I see. Thank you for clarifying 🙂